tools/rust_analyzer/bin/discover_rust_project.rs

```diff
@@ -0,0 +1,265 @@
+    pub bazel: Utf8PathBuf,
+
+    /// The path to a `bazelrc` configuration file.
+    bazelrc: Option<Utf8PathBuf>,
+
+    /// The argument that `rust-analyzer` can pass to the binary.
+    rust_analyzer_argument: Option<RustAnalyzerArg>,
+}
+
+impl Config {
+    // Parse the configuration flags and supplement with bazel info as needed.
+    pub fn parse() -> anyhow::Result<Self> {
+        let ConfigParser {
+            workspace,
+            execution_root,
+            output_base,
+            bazel,
+            bazelrc,
+            rust_analyzer_argument,
+        } = ConfigParser::parse();
+
+        // Implemented this way instead of a classic `if let` to satisfy the
+        // borrow checker.
+        // See: <https://github.com/rust-lang/rust/issues/54663>
+        #[allow(clippy::unnecessary_unwrap)]
+        if workspace.is_some() && execution_root.is_some() && output_base.is_some() {
+            return Ok(Config {
+                workspace: workspace.unwrap(),
+                execution_root: execution_root.unwrap(),
+                output_base: output_base.unwrap(),
+                bazel,
+                bazelrc,
+                rust_analyzer_argument,
+            });
+        }
+
+        // We need some info from `bazel info`. Fetch it now.
+        let mut info_map = get_bazel_info(
+            &bazel,
+            workspace.as_deref(),
+            output_base.as_deref(),
+            bazelrc.as_deref(),
+        )?;
+
+        let config = Config {
+            workspace: info_map
+                .remove("workspace")
+                .expect("'workspace' must exist in bazel info")
+                .into(),
+            execution_root: info_map
+                .remove("execution_root")
+                .expect("'execution_root' must exist in bazel info")
+                .into(),
+            output_base: info_map
+                .remove("output_base")
+                .expect("'output_base' must exist in bazel info")
+                .into(),
+            bazel,
+            bazelrc,
+            rust_analyzer_argument,
+        };
+
+        Ok(config)
+    }
+}
+
+#[derive(Debug, Parser)]
+struct ConfigParser {
+    /// The path to the Bazel workspace directory. If not specified, uses the result of `bazel info workspace`.
+    #[clap(long, env = "BUILD_WORKSPACE_DIRECTORY")]
+    workspace: Option<Utf8PathBuf>,
+
+    /// The path to the Bazel execution root. If not specified, uses the result of `bazel info execution_root`.
+    #[clap(long)]
+    execution_root: Option<Utf8PathBuf>,
+
+    /// The path to the Bazel output user root. If not specified, uses the result of `bazel info output_base`.
+    #[clap(long, env = "OUTPUT_BASE")]
+    output_base: Option<Utf8PathBuf>,
+
+    /// The path to a Bazel binary.
+    #[clap(long, default_value = "bazel")]
+    bazel: Utf8PathBuf,
+
+    /// The path to a `bazelrc` configuration file.
+    #[clap(long)]
+    bazelrc: Option<Utf8PathBuf>,
```

Comment:
as mentioned in the discussion - I'd prefer to leave configurability to a future patch if that's OK, it's not clear whether --bazelrc is the best mechanism - letting the user specify bazel flags might be simpler.